xarray/_array_api/array_object.py

```diff
@@ -0,0 +1,1106 @@
+# Vendored from https://github.com/data-apis/array-api/pull/589
+from __future__ import annotations
+
+from typing import TYPE_CHECKING, Protocol, TypeVar
+
+if TYPE_CHECKING:
+    from xarray._array_api._types import (
+        Any,
+        Enum,
+        PyCapsule,
+        ellipsis,
+    )
+    from xarray._array_api._types import (
+        device as Device,
+    )
+    from xarray._array_api.data_types import DType
+
+array = TypeVar("array", bound="Array")
+# NOTE: when working with py3.11+ this can be ``typing.array``.
+
+
+class Array(Protocol):
```

Comment:
Given that `numpy.array_api` exists, can we replace this entire class with that instead?

Reply:
I don't think that's possible since the types becomes too specific. 
For example `.mT` would always return `numpy.array_api._array_object.Array` which wouldn't work with cubed.

np.array_api also implements a few methods that aren't required by the spec either making it too specific for us.

This code was vendored from a PR in the array-api github so we could hopefully import from there whenever it has been merged.